### PR TITLE
Add InterlockedHL_CompareExchangeIf specs, header, code and tests

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add InterlockedHL_CompareExchangeIf specs, header, code and tests.
Example where this function is useful for switching to a given value (like a faulted state) if we are in any of 2 or more other states.

### CODE REVIEW

File: src/interlocked_hl.c [hunk 1]
```diff
@@ -163,6 +163,53 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForNotVal
...
+            else
+            {
+                /* Codes_SRS_INTERLOCKED_HL_01_014: [ If target changed meanwhile then InterlockedHL_CompareExchangeIf shall return INTERLOCKED_HL_CHANGED and shall not peform any exchange of values. ]*/
+                result = INTERLOCKED_HL_CHANGED;
+            }
+        }
+        else
+        {
+            /* Codes_SRS_INTERLOCKED_HL_01_017: [ If compare returns false then  InterlockedHL_CompareExchangeIf shall not perform any exchanges and return INTERLOCKED_HL_OK. ]*/
+            result = INTERLOCKED_HL_OK;
```
Comment (jebrando) [inline src/interlocked_hl.c:204]: >result = INTERLOCKED_HL_OK; [](http://example.com/codeflow?start=12&length=27)

One thing that makes me think, should we change the return value here to differentiate between a positive compare and a negative compare?

**Review by jebrando — APPROVED** [review]
:shipit:

File: src/interlocked_hl.c [hunk 2]
```diff
@@ -163,6 +163,53 @@ IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForNotVal
     return result;
 }
 
+IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_CompareExchangeIf, int32_t volatile_atomic*, target, int32_t, exchange, INTERLOCKED_COMPARE_EXCHANGE_IF, compare, int32_t*, original_target)
```
Comment (anporumb) [inline src/interlocked_hl.c:166]: >InterlockedHL_CompareExchangeIf [](http://example.com/codeflow?start=53&length=31)

possibly missing "reals" for this new function.

**Review by anporumb — APPROVED** [review]
:shipit:

File: src/interlocked_hl.c (hunk 2 above)
Comment (dcristoloveanu) [reply src/interlocked_hl.c:166]: reals added.

File: src/interlocked_hl.c (hunk 1 above)
Comment (dcristoloveanu) [reply src/interlocked_hl.c:204]: I hear you. Let's leave that for when we hit that case though.

**Pull request merged**